File: javascript/convolution.js
// Function Convolution1D
function Convolution1D(f,g,delay=false)
{
        
        // Define output matrix with zeros
        var y = new Array(f.length+g.length-1).fill(0);
        
        // Convolution
        for (n=0;n<f.length;++n)
        {
            for(m=0;m<g.length;++m)
            {
                
                y[m+n]+=f[n]*g[m];
            }
        }
        if(delay)
        {
              return y.slice(Math.round((g.length-1)/2),-g.length);
        }
	else
	{return y;}
};

module.exports.Convolution1D = Convolution1D;

function Convolution2D(f,g)
{
        // Define output matrix with zeros
        var result = [];
        for ( var i = 0; i<(f.length+g.length-1);++i)
        {
            result[i]=[];
            for (var j = 0 ; j < (f[0].length+g[0].length-1);++j)
            {
             result[i][j]=0;
            }
        }
        
        
        // Convolution
        for (x=0;x<f.length;++x)
        {
            for(y=0;y<f[0].length;++y)
            {
                for(i=0;i<g.length;++i)
                {
                    for (j=0;j<g[0].length;++j)
                    {
                        result[x+i][y+j]+=f[x][y]*g[i][j];
                    }
                }
            }
        }
        
        return result;
      
}

module.exports.Convolution2D = Convolution2D;

File: Cpp/convolution.cpp
#include <vector>
#include <cmath>
#include "convolution.hpp"
typedef std::vector<std::vector<double>> matrix;
using namespace std;

// Define Convolution1D
void Convolution1D (double *f, int fLEN, double *g, int gLEN, double *y)
{
    for (int n=0; n < fLEN; n++)
    {
        for (int m=0; m < gLEN; m++)
        {
            y[n+m] +=  f[n]*g[m];
        }
    }
}

// Define Convolution1D Vector
std::vector<double> Convolution1D_Vec(vector<double> f, vector<double> g, bool delayCorrect)
{
    //Initialize output vector
    std::vector<double> y(f.size()+g.size()-1, 0.0);
    
    
    for(auto fn = f.begin(); fn != f.end();fn++)
    {    
        for(auto gm = g.begin();gm != g.end();gm++)
        {
             y[distance(f.begin(),fn)+distance(g.begin(),gm)]+= (*fn)*(*gm);
        }
    }
    
    if (delayCorrect)
    {
        return vector<double>(y.begin()+ceil((g.size()-1.0)/2.0),y.end()-g.size()+1);
    }
    else
    {
        return y;
    }
       
}

// Define Convolution2D Vector
matrix Convolution2D(matrix &f, matrix &g)
{
    //Initialize output vector
    matrix result(f.size()+g.size()-1, std::vector<double>(f[0].size()+g[0].size()-1,0.0));
    
    
    for(auto fx = f.begin(); fx != f.end();fx++)
    {    
        for(auto fy = fx->begin(); fy != fx->end();fy++)
        {
            for(auto gi = g.begin(); gi != g.end(); gi++)
            {
                for (auto gj = gi->begin(); gj != gi->end(); gj++)
                {
                    result[distance(f.begin(),fx)+distance(g.begin(),gi)][distance(fx->begin(),fy)+distance(gi->begin(),gj)]+= (*fy)*(*gj);
                }
            }
        }
    
    }
    return result;
    
}
